### PR TITLE
use malloc/free for WiFiManagerParameter* array instead of hard coding.

### PR DESCRIPTION
**Pull request by liebman**
Dynamically allocate the _params array.  Existing no arg constructor uses the same default, 10, as it used to.  Added a second constructor that allows setting size.  Added Destructor that frees allocated _params.

### CODE REVIEW

**Comment by merlokk** [comment]
WiFiManager(int max_params = 10);

and return to one constructor

**Comment by tablatronix** [comment]
hmm not sure how I feel about the constructor params

**Comment by liebman** [comment]
Is there a case where constructor params don't work?  Semantically this is no different than multiple constructors and reduces the code complexity.

**Comment by tablatronix** [comment]
I dont like passing params via constructors it causes limitation in the future with regards to backward compatibility

**Comment by liebman** [comment]
I don't think I understand.  Are you saying that there should be no constructors that have parameters?

**Comment by tablatronix** [comment]
not really, just that variable arguments should not be passed via a constructor, a constructor should be passed data that an object needs to work, not a variable parameter that can also be set via setters etc. It present maintenance problems in the future. For example what happens when we refactor the parameter system, or abstract it out into a separate dependency then we have this legacy constructor definition that cannot be reused, and has to be deprecated or handled.

What happens if we need to add more parameters to constructor , what if they are the same datatype ? Its can become a mess.

If we were to say allow a config object to preload parameters, then that would be acceptable.
eg. `constructor(object[arg1:blargh,arg2:blargh]){_localarg1 = checkvaliditiy(object:arg1) }`

**Comment by liebman** [comment]
Ok. I re-worked the change so that it uses the default constructor and resizes the params array each time the size is exceeded. Better?

**Comment by tablatronix** [comment]
Ooh thats a great idea, sorry about this ainwas going to merge and fixup but I have been pre occupied

**Comment by liebman** [comment]
Made the change simpler too.

**Comment by tablatronix** [comment]
I assume you have tested this a bit ?

**Comment by liebman** [comment]
I've been using it in a few projects constantly since the commits.  I don't think we should add limits, if the user wants to add 10,000 params why should we stop them?  ;-)

**Comment by tablatronix** [comment]
Aside from malloc exceptions im sure they will figure it out. Thanks I will merge this then bring it into my dev branch

File: WiFiManager.cpp [hunk 1]
```diff
@@ -60,18 +60,35 @@ const char* WiFiManagerParameter::getCustomHTML() {
...
-	DEBUG_WM(p->getID());
-	return;
+    // rezise the params array
+    _max_params += WIFI_MANAGER_MAX_PARAMS;
+    DEBUG_WM(F("Increasing _max_params to:"));
+    DEBUG_WM(_max_params);
+    WiFiManagerParameter** new_params = (WiFiManagerParameter**)malloc(_max_params * sizeof(WiFiManagerParameter*));
+    // copy old data
+    memcpy(new_params, _params, _paramsCount * sizeof(WiFiManagerParameter*));
+    free(_params);
```
[the remark this answers is not in the document]
Comment (liebman) [reply WiFiManager.cpp]: Very good point.  I'll update the PR

**Comment by tablatronix** [comment]
Changed to == maxparams, and changed max params to 5

**Pull request merged**